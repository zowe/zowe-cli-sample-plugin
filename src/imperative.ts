/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import {IImperativeConfig} from "@brightside/imperative";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    overrides: {
        CredentialManager: "./overrides/CredentialManager.override"
    },
    pluginHealthCheck: __dirname + "/healthCheck.Handler",
    pluginSummary: "Zowe CLI sample plug-in",
    pluginAliases: ["zcsp"],
    rootCommandDescription: "Welcome to the sample plug-in for Zowe CLI!\n\n The sample plug-in " +
        "(& CLI) follows the Zowe CLI command syntax 'zowe [group] [action] [object] [options]'. " +
        "Where, in the case of the plugin, " +
        "the [group] is the package.json name, " +
        "the [actions] are defined in the project under 'src/cli/', " +
        "& the [objects] (definitions & handler) are defined in the project under 'src/cli/'.",
    productDisplayName: "Zowe CLI Sample Plug-in",
    name: "zowe-cli-sample"
};

export = config;
