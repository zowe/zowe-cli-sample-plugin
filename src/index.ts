/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

/**
 * Use the index.ts to export any public facing APIs/intefaces/etc.
 *
 * If your plug-in introduces a set of APIs/functions that others would find useful when building node apps
 * (or editor extensions, etc.) export them here.
 *
 * For example, Zowe CLI offers Jobs APIs that can be invoke programmatically from a VS code extension to create
 * a Job viewer/tree extension.
 */
export * from "./api/Files";
