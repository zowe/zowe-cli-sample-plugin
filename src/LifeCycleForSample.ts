/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import {AbstractPluginLifeCycle, ImperativeError, Logger} from "@zowe/imperative";

/**
 * This class performs lifecycle functions for your plug-in. You are free to
 * change the name of this class to reflect your plug-in and its actions.
 *
 * You only have to supply such a class if you specify the 'pluginLifeCycle'
 * property in your IImperativeConfig definition for your plug-in. In this sample,
 * that definition is supplied in src/pluginDef.ts.
 */
export class LifeCycleForSample extends AbstractPluginLifeCycle {
    /**
     * This function will be called after a plug-in is installed.
     * A plug-in can use this opportunity to perform a sanity test or to
     * perform some additional setup which is specific to your plug-in.
     *
     * If your plug-in's postInstall function performs an asynchronous operation,
     * its postInstall function should return a promise, allowing the CLI
     * to await the postInstall function. Such a signature would be like this:
     *
     *      public async postInstall(): Promise<void>
     *
     * If it has no asynchronous operation, the postInstall function can be
     * written as a synchronous function. That is the style which this trivial
     * synchronous sample uses below.
     *
     * If you supply this class, you must supply this function. If you have nothing to
     * do post-install, then simply return from the function.
     */
    public postInstall(): void {
        /* It is up to you to decide what actions are meaningful for you
         * to perform immediately after your plug-in is installed.
         *
         * When all goes well, you probably should display nothing to the user.
         */

        const somethingWeirdOccurred = true;
        if ( somethingWeirdOccurred ) {
            /* If you display a message about a problem, you should also log the
             * message to a file, so that the message can be found later.
             *
             * You can use a consoleLogger to display a message. The available logging
             * levels can be found in the Zowe doc:
             * https://docs.zowe.org/stable/user-guide/cli-configuringcli-ev/#setting-cli-log-levels
             *
             * Note that the user controls the logging level with the
             * ZOWE_IMPERATIVE_LOG_LEVEL environment variable. When not set, the
             * default log level is WARN. So by default, only messages displayed at
             * the WARN, ERROR, and FATAL levels will be displayed (or logged).
             */
            const warnMsg = "Something weird happened in the sample plug-in's " +
                 "postInstall function. However, things should still work ok.";
            Logger.getConsoleLogger().warn(warnMsg);
            Logger.getImperativeLogger().warn(warnMsg);
        }

        /* When you are unable to achieve your desired actions, you should throw
         * an imperative error. The "zowe plugins install" command will catch
         * that error, display it, and add some additional information.
         *
         * Your plug-in has already been installed. However, when you throw an
         * error during your postInstall function, the 'zowe plugins install'
         * command will exit with a non-zero exit code.
         */
        const someBigProblemOccurred = false;
        if ( someBigProblemOccurred ) {
            throw new ImperativeError({
                msg: "Something awful happened in the sample plug-in's " +
                     "postInstall function. Thus, the plug-in could not be properly setup."
            });
        }
    }

    /**
     * This function will be called before a plug-in is uninstalled.
     * This lifecycle hook is intended to replace the capability that used to
     * be performed by the NPM pre-uninstall action before NPM removed that
     * capability in NPM version 7.
     * See https://docs.npmjs.com/cli/v9/using-npm/scripts#a-note-on-a-lack-of-npm-uninstall-scripts
     *
     * A plug-in can use this opportunity to revert any specialized setup that was
     * established during the lifetime of your plug-in.
     *
     * If your plug-in's preUninstall function performs an asynchronous operation,
     * its preUninstall function should return a promise, allowing the CLI
     * framework to await the preUninstall function. Such a signature would be like this:
     *
     *      public async preUninstall(): Promise<void>
     *
     * If it has no asynchronous operation, the preUninstall function can be
     * written as a synchronous function. That is the style which this trivial
     * synchronous sample uses below.
     *
     * If you supply this class, you must supply this function. If you have nothing to
     * do pre-uninstall, then simply return from the function.
     */
    public preUninstall(): void {
        /* The comments from the postInstall function also apply to the
         * preUninstall function.
         */
        const somethingWeirdOccurred = true;
        if ( somethingWeirdOccurred ) {
            const warnMsg = "Something weird happened in the sample plug-in's " +
                "preUninstall function that we thought you should know about.";
            Logger.getConsoleLogger().warn(warnMsg);
            Logger.getImperativeLogger().warn(warnMsg);
        }

        /* When you throw an error during your preUninstall function, the
         * 'zowe plugins uninstall' catches the error and still performs the
         * uninstall operation that was requested by the user. If the actual
         * uninstall operation succeeds, the command will exit with a zero
         * exit code.
         */
        const someBigProblemOccurred = false;
        if ( someBigProblemOccurred ) {
            throw new ImperativeError({
                msg: "Something awful happened during the sample plug-in's " +
                     "preUninstall function. Maybe you have to manually release " +
                     "some resource that was used by this plug-in."
            });
        }
    }
}

module.exports = LifeCycleForSample;