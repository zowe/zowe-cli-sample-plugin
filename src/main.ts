#!/usr/bin/env node
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/


import { Imperative } from "@brightside/imperative";
// init imperative & parse the CLI input.
Imperative.init().then(() => {
    Imperative.api.imperativeLogger.info("Imperative Initialized the sample plugin CLI!");
    Imperative.parse();
    Imperative.api.imperativeLogger.info("Parsed the input command!");
}).catch((error) => {
    Imperative.console.error(`An error occured during imperative initalization:\n${error}`);
});
