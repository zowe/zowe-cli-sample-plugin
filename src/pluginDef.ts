/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import {IImperativeConfig} from "@zowe/imperative";
import { ProfileConstants } from "./api/ProfileConstants";

const pluginDef: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    pluginSummary: "Zowe CLI sample plug-in",
    pluginAliases: ["zcsp"],
    rootCommandDescription: "Welcome to the sample plug-in for Zowe CLI!\n\n The sample plug-in " +
        "(& CLI) follows the Zowe CLI command syntax 'zowe [group] [action] [object] [options]'. " +
        "Where, in the case of the plugin, " +
        "the [group] is the package.json name, " +
        "the [actions] are defined in the project under 'src/cli/', " +
        "& the [objects] (definitions & handler) are defined in the project under 'src/cli/'.",
    productDisplayName: "Zowe CLI Sample Plug-in",
    name: "zowe-cli-sample",
    profiles: [
        {
            type: "sample",
            schema: {
                type: "object",
                title: "Sample Profile",
                description: "A sample profile is required to issue commands that interact with the sample API. The " +
                    "sample profile contains your host, port, user name, and password to connect to the sample API.",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: ProfileConstants.SAMPLE_OPTION_HOST
                    },
                    port: {
                        type: "number",
                        includeInTemplate: true,
                        optionDefinition: ProfileConstants.SAMPLE_OPTION_PORT
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: ProfileConstants.SAMPLE_OPTION_USER
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: ProfileConstants.SAMPLE_OPTION_PASSWORD
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: ProfileConstants.SAMPLE_OPTION_REJECT_UNAUTHORIZED
                    },
                    tshirtSize: {
                        type: "string",
                        optionDefinition: ProfileConstants.SAMPLE_OPTION_TSHIRT_SIZE
                    }
                }
            }
        }
    ]
};

export = pluginDef;
