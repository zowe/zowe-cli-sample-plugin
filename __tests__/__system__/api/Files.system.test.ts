/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import {Files} from "../../../src/index";

describe("Files", () => {
    it ("should read the example directory", () => {
        // Get the example directory contents
        const contents = Files.listDirectoryContents(__dirname + "/../../__resources__/example_directory/");

        // Expect the contents to exist & have two entries... and expect that the fields are present
        expect(contents).toBeDefined();
        expect(contents.length).toBe(2);
        contents.forEach((entry) => {
            expect(entry.mode).toBeDefined();
            expect(entry.size).toBeDefined();
            expect(entry.birthed).toBeDefined();
            expect(entry.lastModified).toBeDefined();
            expect(entry.name).toMatchSnapshot();
        });
    });
});
