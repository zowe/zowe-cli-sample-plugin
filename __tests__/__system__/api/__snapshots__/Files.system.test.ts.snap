// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files should read the example directory 1`] = `"ExampleFile.txt"`;

exports[`Files should read the example directory 2`] = `"ExampleFolder"`;
